File: 11/1.swift
import Foundation

enum InputOption {
	case test
	case partOne
	case partTwo
}

let elems: [String]
let floorMasks: [Int]
let runOption = InputOption.partOne  // Change this to try different inputs.
switch runOption {
	case .test: 
		// .  .  .  .
		// .  .  LG .
		// HG .  .  .
		// .  HM .  LM
		elems = ["H", "L"]
		floorMasks = [0b0101, 0b1000, 0b0010, 0b0000]
	case .partOne:
		// .  .  .  .  .  .  .  .  .  .
		// .  .  .  .  .  .  XG XM RG RM
		// .  .  .  PM .  SM .  .  .  .
		// TG TM PG .  SG .  .  .  .  .
		elems = ["T", "P", "S", "X", "R"]
		floorMasks = [0b1110100000, 0b0001010000, 0b0000001111, 0b0000000000]
	case .partTwo:
		// F4 .  .  .  .  .  .  .  .  .  .  .  .  .  .  .  
		// F3 .  .  .  .  .  .  .  XG XM RG RM .  .  .  .  
		// F2 .  .  .  .  PM .  SM .  .  .  .  .  .  .  .  
		// F1 E  TG TM PG .  SG .  .  .  .  .  EG EM DG DM 
		elems = ["T", "P", "S", "X", "R", "E", "D"]
		floorMasks = [0b11101000001111, 0b00010100000000, 0b00000011110000, 0b0000000000]
}

extension Int {
	func brief(_ digits: Int? = nil) -> String {
		let digs = digits == nil ? 2*elems.count : digits!
		var s = ""
		for pos in Swift.stride(from: digs - 1, through: 0, by: -1) {
			s += (self & (1<<pos) == 0 ? "0" : "1")
		}
		return s
	}
}

struct Group: CustomStringConvertible {
	let mask: Int  // bit width is 2*elems.count

	var brief: String {
		return mask.brief()
	}

	var isDeadly: Bool {
		return containsUnprotectedMicrochip && containsGenerator
	}

	var containsUnprotectedMicrochip: Bool {
		for pos in 0..<elems.count {
			if mask & 1<<(2*pos) != 0 && mask & 1<<(2*pos+1) == 0 {
				return true
			}
		}
		return false
	}

	var containsGenerator: Bool {
		for pos in 0..<elems.count {
			if mask & 1<<(2*pos+1) != 0 {
				return true
			}
		}
		return false
	}

	var description: String {
		var d = ""
		var m = mask
		for elemLetter in elems.reversed() {
			let twoBits = m & 0b11
			d = (twoBits & 0b01 == 0 ? ".  " : "\(elemLetter)M ") + d
			d = (twoBits & 0b10 == 0 ? ".  " : "\(elemLetter)G ") + d
			m = m>>2
		}
		return d
	}

	func possibleLoadMasks() -> [Int] {
		var loads: [Int] = []
		for pos in 0..<2*elems.count {
			if mask & 1<<pos != 0 {
				loads.append(1<<pos)
			}
		}
		let numSingleLoads = loads.count
		for i in 0..<numSingleLoads-1 {
			for j in i+1..<numSingleLoads {
				loads.append(loads[i] | loads[j])
			}
		}
		return loads
	}
}

class Building: CustomStringConvertible {
	let floors: [Group]
	let elevator: Int

	var brief: String {
		return elevator.brief(2) + floors.map({ $0.brief }).joined()
	}

	var mask: Int {
		var m = elevator
		for f in floors {
			m <<= 2*elems.count
			m |= f.mask
		}
		return m
	}

	var isDeadly: Bool {
		for f in floors {
			if f.isDeadly {
				return true
			}
		}
		return false
	}

	var missionAccomplished: Bool {
		return floors.last!.mask == 1<<(2*elems.count) - 1
	}

	init(_ floors: [Group], _ elevator: Int) {
		self.floors = floors
		self.elevator = elevator
		Building.sanityCheck(floors, elevator)
	}

	convenience init(_ floorMasks: [Int], _ elevator: Int) {
		self.init(floorMasks.map({ Group(mask: $0) }), elevator)
	}

	var description: String {
		var d = ""
		for floorIndex in stride(from: floors.count - 1, through: 0, by: -1) {
			d += "F\(floorIndex + 1) "
			if floorIndex == elevator {
				d += "E  "
			} else {
				d += ".  "
			}
			d += "\(floors[floorIndex])\n"
		}
		return d
	}

	func possibleLoadMasks() -> [Int] {
		return floors[elevator].possibleLoadMasks()
	}

	func resultOfMove(loadMask: Int, direction: Int) -> Building? {
		let newElevator = elevator + direction
		if newElevator < 0 || newElevator >= floors.count {
			return nil
		}

		var newFloors = floors
		newFloors[newElevator] = Group(mask: floors[newElevator].mask | loadMask)
		newFloors[elevator] = Group(mask: floors[elevator].mask & ~loadMask)

		let newBuilding = Building(newFloors, newElevator)
		return newBuilding
	}

	private static func sanityCheck(_ floors: [Group], _ elevator: Int) {
		var combined = 0
		let bitwidth = 2 * elems.count
		for f in floors {
			if f.mask >= 1 << bitwidth {
				fatalError("floor mask 0b\(f.brief) has too many bits")
			}
			if f.mask & combined != 0 {
				fatalError("floor mask 0b\(f.brief) contains a bit already used")
			}
			combined |= f.mask
		}
		if combined != (1 << bitwidth) - 1 {
			let formatString = "%\(bitwidth * floors.count)b"
			fatalError("combined mask 0b\(String(format: formatString, combined)) is missing a bit")
		}

		if !(0..<floors.count).contains(elevator) {
			fatalError("elevator \(elevator) is out of bounds")
		}
	}
}

class Move {
	let depth: Int
	let prev: Move?
	let resultOfMove: Building

	init(depth: Int, prev: Move?, resultOfMove: Building) {
		self.depth = depth
		self.prev = prev
		self.resultOfMove = resultOfMove
	}
}

class QueueNode<T> {
	let value: T
	var next: QueueNode<T>?
	init(value: T) {
		self.value = value
	}
}

struct Queue<T> {
	private(set) var count: Int = 0
	private var head: QueueNode<T>?
	private var tail: QueueNode<T>?

	mutating func push(_ value: T) {
		let node = QueueNode<T>(value: value)
		if let oldTail = tail {
			oldTail.next = node
			tail = node
		} else {
			tail = node
			head = node
		}
		count += 1
	}

	mutating func pop() -> T? {
		if let oldHead = head {
			head = oldHead.next
			if head == nil {
				tail = nil
			}
			count -= 1
			return oldHead.value
		} else {
			return nil
		}
	}
}

protocol MoveQueue {
	var count: Int { get }
	mutating func push(_: Move)
	mutating func pop() -> Move?
}

struct MoveQueueUsingArray: MoveQueue {
	private var moves: [Move] = []

	var count: Int {
		return moves.count
	}

	mutating func push(_ move: Move) {
		moves.append(move)
	}
	
	mutating func pop() -> Move? {
		return moves.isEmpty ? nil : moves.removeFirst()
	}
}

struct MoveQueueUsingLinkedList: MoveQueue {
	private var moves = Queue<Move>()

	var count: Int {
		return moves.count
	}

	mutating func push(_ move: Move) {
		moves.push(move)
	}
	
	mutating func pop() -> Move? {
		return moves.pop()
	}
}


func doSearch(usingLinkedList: Bool) {
	NSLog("starting search using %@ for the BFS queue", usingLinkedList ? "linked list" : "array")

	let building = Building(floorMasks, 0)
	print(building)

	let depthCheckpointFactor = 25
	var lastCheckpoint = 50
	var visitedBuildingMasks = Set<Int>()

	var queue: MoveQueue = usingLinkedList ? MoveQueueUsingLinkedList() : MoveQueueUsingArray()
	queue.push(Move(depth: 0, prev: nil, resultOfMove: building))

	var maxQueueLength = queue.count
	while queue.count > 0 {
		let move = queue.pop()!
		if move.depth >= lastCheckpoint && move.depth % depthCheckpointFactor == 0 {
			print("about to go to depth \(move.depth), queue length is \(queue.count)")
			lastCheckpoint += depthCheckpointFactor
		}
		if move.resultOfMove.missionAccomplished {
			NSLog("Mission accomplished at depth %ld", move.depth)
			NSLog("maxQueueLength: %ld, maxDepth: %ld, visited: %ld", maxQueueLength, visitedBuildingMasks.count)
			
			// Note: this displays the moves in reverse order.
//			var m = move
//			while m.prev != nil {
//				print(m.resultOfMove)
//				m = m.prev!
//			}
			break
		}
		let possibleLoadMasks = move.resultOfMove.possibleLoadMasks()
		for loadMask in possibleLoadMasks {
			for dir in [-1, 1] {
				if let b2 = move.resultOfMove.resultOfMove(loadMask: loadMask, direction: dir) {
					let b2mask = b2.mask
					if visitedBuildingMasks.contains(b2mask) {
						continue
					}
					visitedBuildingMasks.insert(b2mask)
					if visitedBuildingMasks.count % 5000000 == 0 {
						NSLog("visited: %ld", visitedBuildingMasks.count)
					}
					if b2.isDeadly {
						continue
					}
					queue.push(Move(depth: move.depth+1, prev: move, resultOfMove: b2))
					if maxQueueLength < queue.count {
						maxQueueLength = queue.count
						if maxQueueLength % 50000 == 0 {
							NSLog("maxQueueLength: %ld", maxQueueLength)
						}
					}
				}
			}
		}
	}
}

doSearch(usingLinkedList: false)
doSearch(usingLinkedList: true)


//print(Group(mask: 0))
//print(Group(mask: 0b0010))
//print(Group(mask: 0b1000))
//print(Group(mask: 0b0101))
//

//print(Group(mask: 0b0101).isDeadly)
//print(Group(mask: 0b0101).containsGenerator)
//print(Group(mask: 0b0101).containsUnprotectedMicrochip)
//print(Group(mask: 0b0111).possibleLoadMasks())

File: 11/1.py
from copy import deepcopy
import datetime

(g_elements, g_floorMasks) = {
	111: (["H", "L"], [0b0101, 0b1000, 0b0010, 0b0000]),
	222: (["T", "P", "S", "X", "R"], [0b1110100000, 0b0001010000, 0b0000001111, 0b0000000000]),
	333: (["T", "P", "S", "X", "R", "E", "D"],
			[0b11101000001111, 0b00010100000000, 0b00000011110000, 0b0000000000])
}[222]

g_bitsPerFloor = 2*len(g_elements)

def fatalError(message):
	print("FATAL ERROR: " + message)
	exit

class Group(object):
	def __init__(self, mask):
		self.mask = mask

	@property
	def isDeadly(self):
		return self.containsUnprotectedMicrochip and self.containsGenerator

	@property
 	def containsUnprotectedMicrochip(self):
		for elementIndex in range(0, len(g_elements)):
			if self.mask & 1<<(2*elementIndex) != 0 and self.mask & 1<<(2*elementIndex+1) == 0:
				return True
		return False

	@property
	def containsGenerator(self):
		for elementIndex in range(0, len(g_elements)):
			if self.mask & 1<<(2*elementIndex+1) != 0:
				return True
		return False

	@property
	def possibleLoadMasks(self):
		loads = []
		# All possible single loads.
		for pos in range(0, g_bitsPerFloor):
			if self.mask & 1<<pos != 0:
				loads.append(1<<pos)
		# All possible double loads.
		numSingleLoads = len(loads)
		for i in range(0, numSingleLoads - 1):  # i ranges from first to next-to-last
			for j in range(i + 1, numSingleLoads):  # j ranges from i+1 to last
				loads.append(loads[i] | loads[j])
		return loads

	@property
	def description(self):
		d = ""
		m = self.mask
		for elemLetter in reversed(g_elements):
			lastTwoBits = m & 0b11
			d = (".  " if (lastTwoBits & 0b01 == 0) else "%sM " % elemLetter) + d
			d = (".  " if (lastTwoBits & 0b10 == 0) else "%sG " % elemLetter) + d
			m >>= 2
		return d

class Building(object):
	def __init__(self, floors, elevator):
		self.sanityCheck(floors, elevator)
		self.floors = floors
		self.elevator = elevator

	@property
	def mask(self):
		m = self.elevator
		for f in self.floors:
			m <<= g_bitsPerFloor
			m |= f.mask
		return m

	@property
	def isDeadly(self):
		for f in self.floors:
			if f.isDeadly:
				return True
		return False

	@property
	def missionAccomplished(self):
		return self.floors[-1].mask == (1<<g_bitsPerFloor) - 1

	@property
	def possibleLoadMasks(self):
		return self.floors[self.elevator].possibleLoadMasks

	def resultOfMove(self, loadMask, direction):
		newElevator = self.elevator + direction
		if newElevator < 0 or newElevator >= len(self.floors):
			return None

		newFloors = deepcopy(self.floors)
		newFloors[newElevator] = Group(self.floors[newElevator].mask | loadMask)
		newFloors[self.elevator] = Group(self.floors[self.elevator].mask & ~loadMask)

		newBuilding = Building(floors = newFloors, elevator = newElevator)
		return newBuilding

	def sanityCheck(self, floors, elevator):
		combined = 0
		for f in floors:
			if f.mask >= (1 << g_bitsPerFloor):
				fatalError("floor mask 0b\(f.brief) has too many bits")
			if f.mask & combined != 0:
				fatalError("floor mask 0b\(f.brief) contains a bit already used")
			combined |= f.mask

		if combined != (1 << g_bitsPerFloor) - 1:
			formatString = "%\(g_bitsPerFloor * floors.count)b"
			fatalError("combined mask 0b\(String(format: formatString, combined)) is missing a bit")

		if not elevator in range(0, len(floors)):
			fatalError("elevator \(elevator) is out of bounds")

	@property
	def description(self):
		d = ""
		for floorIndex in range(len(self.floors) - 1, -1, -1):
			d += "F{} ".format(floorIndex + 1)
			d += "E  " if floorIndex == self.elevator else ".  "
			d += self.floors[floorIndex].description
			d += "\n"
		return d

class Move(object):
	def __init__(self, depth, prevMove, resultingBuilding):
		self.depth = depth
		self.prevMove = prevMove
		self.resultingBuilding = resultingBuilding

	def dump(self):
		moves = []
		m = self
		while m:
			moves.append(m)
			m = m.prevMove
		for mv in reversed(moves):
			print(mv.resultingBuilding.description)

class QueueNode(object):
	def __init__(self, value):
		self.value = value
		self.next = None

class Queue(object):
	def __init__(self):
		self.count = 0
		self.head = None
		self.tail = None

	def push(self, value):
		node = QueueNode(value)
		oldTail = self.tail
		if oldTail:
			oldTail.next = node
			self.tail = node
		else:
			self.tail = node
			self.head = node
		self.count += 1

	def pop(self):
		oldHead = self.head
		if oldHead:
			self.head = oldHead.next
			if self.head is None:
				self.tail = None
			self.count -= 1
			return oldHead.value
		else:
			return None

def doSearch(usingLinkedList):
	#NSLog("starting search using %@ for the BFS queue", usingLinkedList ? "linked list" : "array")

	building = Building(map(lambda x: Group(x), g_floorMasks), 0)
#	print("BUILDING:")
#	print(building.description)

	print("[{}] START".format(datetime.datetime.now()))
	depthCheckpointFactor = 25
	lastCheckpoint = 50
	visitedBuildingMasks = set()

	queue = Queue()
	queue.push(Move(0, None, building))

	maxQueueLength = queue.count
	while queue.count > 0:
		move = queue.pop()
		if (move.depth >= lastCheckpoint) and (move.depth % depthCheckpointFactor == 0):
			print("About to go to depth {}, queue length is {}".format(move.depth, queue.count))
			lastCheckpoint += depthCheckpointFactor
		if move.resultingBuilding.missionAccomplished:
			print("[{}] END at depth {}".format(datetime.datetime.now(), move.depth))
			print("maxQueueLength: {}, visited: {}".format(maxQueueLength, len(visitedBuildingMasks)))
			#move.dump()
			print("")
			break
		possibleLoadMasks = move.resultingBuilding.possibleLoadMasks
		for loadMask in possibleLoadMasks:
			for dir in [-1, 1]:
				b2 = move.resultingBuilding.resultOfMove(loadMask, dir)
				if b2:
					b2mask = b2.mask
					if b2mask in visitedBuildingMasks:
						continue
					visitedBuildingMasks.add(b2mask)
					if len(visitedBuildingMasks) % 5000000 == 0:
						print("visited: {}".format(len(visitedBuildingMasks)))
					if b2.isDeadly:
						continue
					queue.push(Move(move.depth+1, move, b2))
					if maxQueueLength < queue.count:
						maxQueueLength = queue.count
						if maxQueueLength % 50000 == 0:
							print("maxQueueLength: {}".format(maxQueueLength))
	print("DONE -- %s" % datetime.datetime.now())

#doSearch(usingLinkedList = False)
doSearch(usingLinkedList = True)


#print(Group(mask: 0))
#print(Group(mask: 0b0010))
#print(Group(mask: 0b1000))
#print(Group(mask: 0b0101))
#

#print(Group(mask: 0b0101).isDeadly)
#print(Group(mask: 0b0101).containsGenerator)
#print(Group(mask: 0b0101).containsUnprotectedMicrochip)
#print(Group(mask: 0b0111).possibleLoadMasks())




